### ISSUE TITLE
Changes detected but not tracked on Windows 10 and possibly other platforms

### ISSUE THREAD

**Issue by ctxnop** [syncthing/syncthing-inotify#164]
## The problem
When I start syncthing-inotify and immediately modify a synced file, the change is detected and the sync occurs. But if I wait before making a change, or make a second change, then nothing happen. Sometimes changes are detected and tracked but after a long delay, and then no more changes are tracked.

If I start whith -verbosity=4 I see that the change is detected, but not tracked.

I have this problem every time on Windows, but a coworker of mine did had this problem once on Linux but we can't manage to reproduce every time on Linux.

## How to reproduce
* Start SyncThing
* Start SyncThing-Inotify with -verbosity=4
* Do a change in a tracked folder (create, delete or edit a file)
* See in the console that the change is detected and tracked
* Do another change in the tracked folder
* See in the console that the change is detected, but not tracked.

## Debugging
I don't know Go at all, but I give a try to look into the code anyway, so I may be completely wrong on the reasons.

The problem seem to occurs [on the line 680 of the syncwatcher.go file](https://github.com/syncthing/syncthing-inotify/blob/master/syncwatcher.go#L680)

The code is :
```go
if flushTimerNeedsReset {
	flushTimerNeedsReset = false
	if !flushTimer.Stop() {
		<-flushTimer.C // The problematic line
	}
	flushTimer.Reset(currInterval)
}
```
This line never return, like if there is no event on the channel.
The reason seem to be that, [according to documentation](https://golang.org/pkg/time/#Ticker.Stop), you cannot do this in a concurrent context.
So there is a problem because the `select` which is right after have this `case <-flushTimer.C:`.

## Workaround
I tried a dirty workaround by making these two changes (see the patch in attachment)

```go
	timerStopped := false // Add a new boolean variable
	for {
		if flushTimerNeedsReset {
			flushTimerNeedsReset = false
			//if !flushTimer.Stop() {
			//	<-flushTimer.C
			//}
			timerStopped = true // "stop" the timer
			flushTimer.Stop()
			flushTimer.Reset(currInterval)
			timerStopped = true // "unstop" the timer"
		}
```
and 
```go
		case <-flushTimer.C:
			if timerStopped == true { // Check the "stop" status of the timer
				continue
			}
```

I'm pretty sure this can bring other kind of race conditions, like I said, I know nothing about Go. But using this workaround, every change are detected and tracked right away.

## Patch file
```diff
--- a/syncwatcher.go	2017-05-05 14:15:22.652212700 +0200
+++ b/syncwatcher.go	2017-05-05 14:22:17.755737900 +0200
@@ -673,13 +673,17 @@
 	nextScanTime := time.Now().Add(delayScanInterval) // Time to remind Syncthing to delay scan
 	flushTimer := time.NewTimer(0)
 	flushTimerNeedsReset := true
+	timerStopped := false
 	for {
 		if flushTimerNeedsReset {
 			flushTimerNeedsReset = false
-			if !flushTimer.Stop() {
-				<-flushTimer.C
-			}
+			//if !flushTimer.Stop() {
+			//	<-flushTimer.C
+			//}
+			timerStopped = true
+			flushTimer.Stop()
 			flushTimer.Reset(currInterval)
+			timerStopped = true
 		}
 		select {
 		case item := <-stInput:
@@ -721,6 +725,9 @@
 			Debug.Println("[FS] Tracking: " + item)
 			inProgress[item] = progressTime{true, time.Now()}
 		case <-flushTimer.C:
+			if timerStopped == true {
+				continue
+			}
 			flushTimerNeedsReset = true
 			if delayScan > 0 && nextScanTime.Before(time.Now()) {
 				nextScanTime = time.Now().Add(delayScanInterval)
```

**Comment by AudriusButkevicius** [comment]
The right solution here would be to replace the timer read with an optional read.

```
select {
  case <-flushTimer.C:
  default:
}
```

**Comment by ctxnop** [comment]
Your solution seem to work but there is two things happening, don't know if normal or not.
Here is my console ouput of the `syncthing-inotify -verbority=4` when I just write a line into a file called "seb.txt".

```
[DEBUG] 16:13:36 Empty paths
[DEBUG] 16:13:36 Timeout AccumulateChanges
[DEBUG] 16:13:36 Waiting for seb.txt
[DEBUG] 16:13:36 Empty paths
[DEBUG] 16:13:36 Timeout AccumulateChanges
[DEBUG] 16:13:36 Waiting for seb.txt
[DEBUG] 16:13:36 Empty paths
[DEBUG] 16:13:36 Timeout AccumulateChanges
[DEBUG] 16:13:36 Waiting for seb.txt
[DEBUG] 16:13:36 Empty paths
[DEBUG] 16:13:36 Timeout AccumulateChanges
[DEBUG] 16:13:36 Informing about seb.txt
[DEBUG] 16:13:36 [AG] Not found: seb.txt
[DEBUG] 16:13:36 [AG] Appending path: seb.txt seb.txt
[TRACE] 16:13:36 Informing ST: CKI7R47-UWNDQC3_agent-TCF: [seb.txt]
[TRACE] 16:13:36 Requesting STEvents: 81
[OK] 16:13:36 Syncthing is indexing change in CKI7R47-UWNDQC3_agent-TCF: [seb.txt]
[DEBUG] 16:13:36 [INFORMED] Removed tracking for seb.txt
[DEBUG] 16:13:36 Slowing down inotify timeout parameters for CKI7R47-UWNDQC3_agent-TCF
[TRACE] 16:13:36 Requesting STEvents: 83
[TRACE] 16:13:36 Requesting STEvents: 84
[DEBUG] 16:13:36 [ST] Incoming Changes for CKI7R47-UWNDQC3_agent-TCF, speeding up inotify timeout parameters to 500ms
[DEBUG] 16:13:36 Slowing down inotify timeout parameters for CKI7R47-UWNDQC3_agent-TCF
[TRACE] 16:13:36 Requesting STEvents: 85
[TRACE] 16:13:36 Requesting STEvents: 86
[TRACE] 16:13:36 Requesting STEvents: 87
[TRACE] 16:13:36 Requesting STEvents: 88
[TRACE] 16:14:36 Requesting STEvents: 88
[TRACE] 16:15:36 Requesting STEvents: 88
```

1st, there is a delay of 2-10 seconds, between edition and detection. Not too problematic, but still.

2nd, the 3 first line are repeated almost 100 000 times! (my log file, after a single change, is 92 536 long).

The first two problem didn't occurs with my patch, here the output of the same command, same file edition, but with my patch:

```
[TRACE] 16:20:03 Testing WebGUI
[TRACE] 16:20:03 Getting Folders
[DEBUG] 16:20:03 Installing watch for Project_agent-TCF
[TRACE] 16:20:03 Requesting STEvents: 0
[TRACE] 16:20:03 Getting ignore patterns for Project_agent-TCF
[OK] 16:20:03 Watching Project_agent-TCF: C:\Users\ctxnop\SyncThing
[DEBUG] 16:20:03 Delay scan reminder interval for CKI7R47-UWNDQC3_agent-TCF set to 3595 seconds
[TRACE] 16:20:03 Asking to delay full scanning of CKI7R47-UWNDQC3_agent-TCF
[TRACE] 16:20:03 Informing ST: CKI7R47-UWNDQC3_agent-TCF: [.stfolder]
[OK] 16:20:03 Syncthing is indexing change in CKI7R47-UWNDQC3_agent-TCF: [.stfolder]
[DEBUG] 16:20:03 [ST] Incoming Changes for CKI7R47-UWNDQC3_agent-TCF, speeding up inotify timeout parameters to 500ms
[DEBUG] 16:20:03 [ST] Incoming Changes for CKI7R47-UWNDQC3_agent-TCF
[DEBUG] 16:20:03 [ST] Incoming Changes for CKI7R47-UWNDQC3_agent-TCF
[DEBUG] 16:20:03 [ST] Incoming Changes for CKI7R47-UWNDQC3_agent-TCF
[DEBUG] 16:20:03 [ST] Incoming Changes for CKI7R47-UWNDQC3_agent-TCF
[DEBUG] 16:20:03 [ST] Incoming Changes for CKI7R47-UWNDQC3_agent-TCF
[TRACE] 16:20:03 Requesting STEvents: 88
[TRACE] 16:20:03 Requesting STEvents: 90
[DEBUG] 16:20:27 Change detected in: C:\Users\ctxnop\SyncThing\seb.txt (could still be ignored)
[TRACE] 16:20:27 Change detected in: C:\Users\ctxnop\SyncThing\seb.txt
[DEBUG] 16:20:27 [FS] Incoming Changes for CKI7R47-UWNDQC3_agent-TCF
[DEBUG] 16:20:27 [FS] Tracking: seb.txt
[DEBUG] 16:20:27 Change detected in: C:\Users\ctxnop\SyncThing\seb.txt (could still be ignored)
[TRACE] 16:20:27 Change detected in: C:\Users\ctxnop\SyncThing\seb.txt
[DEBUG] 16:20:27 [FS] Incoming Changes for CKI7R47-UWNDQC3_agent-TCF
[DEBUG] 16:20:27 [FS] Tracking: seb.txt
[TRACE] 16:21:03 Requesting STEvents: 90
[TRACE] 16:22:03 Requesting STEvents: 90
[TRACE] 16:23:03 Requesting STEvents: 90
```
As you may notice, I currently use a polling of 60s in syncthing.

**Comment by imsodin** [comment]
The timer `flushTimer` is local to the `accumulatesChanges` function, which doesn't do anything concurrently. So stopping the timer the way it is should work fine.
What made you think this part is responsible for your problems in the first place?

**Comment by ctxnop** [comment]
By following logs it's clear the problem occurs in this function. Because the [last log](https://github.com/syncthing/syncthing-inotify/blob/master/syncwatcher.go#L680) comes from function which write into the `fsInput` channel.

So the change is detected, processed with ignoring filter, then the path is written to the fsInput channel. Then, if the change is not committed, the problem must be in the `accumulatesChanges`, which is started with the `go` keyword, so it's a thread if I understand well.

At first I thought there is some kind of exception which exit the thread for whaterver reason, but debugging shows that the go routine was still alive when I tried to change some files which were not tracked.

Then I put some more log to see what happen and it appears that, like I said, the line `<-flushTimer.C` never return because my last log was the line I put right before this one. After waiting some times it was still frozen in this line.

I thought there was an algebra error when calculating the timer duration, bug debugging show reasonable time, and it wouldn't explain that the fsInput case stop being triggered. The only reason is the block isn't in the select block. So it must be either before or into a case, but my last log was before the `<-flushTimer.C` line.

To confirm that the problem come with this timer, I simply commented out the following code
```go
if flushTimerNeedsReset {
	flushTimerNeedsReset = false
	if !flushTimer.Stop() {
		<-flushTimer.C
	}
	flushTimer.Reset(currInterval)
}
```
This change confirm that the `<-flushTimer.C` was the problem because now everything goes fine. Changes are detected and tracked, changes are applied on the sync destination.
Of course, it's not a good change because it completely disable the flushTimer which can have some side effects. That's why I came to my patch, so that the timer isn't disabled, but the the problematic construction shown in the documentation is gone. Then everything was still fine. I let it run almost one hour with some changes to sync from time to time, everything seems to went fine.

Are you sure this function doesn't do anything concurrently ? I mean, the select case is in this form
```go
select {
    case <-someChannel:
        // ...
    case <-someOtherChannel:
        // ...
    case <-timerChannel:
        // ...
}
```
As I said, I'm discovering the go language, but as I understand it, because reading a channel is a blocking operation, this must be a concurrent construction in which 3 cases are waiting concurrently, the first channel to receive data will unblock the flow by entering into the corresponding case.
How can you be waiting 3 different channels simultaneously without being in a concurrent context ?

If i's not concurrent, then it will that the flow block on the reading of the first channel, waiting for some data and it will considering reading the second channel only after something is read from the first.
I'm pretty sure that it's not what it's wanted here, because it will be very inefficient.

**Comment by imsodin** [comment]
Sure, in that sense it is concurrent while waiting. The "flow" however isn't, because as you said only first case to receive something is executed, never two cases concurrently. Can you post some logs, maybe even with your extra debug statements.

I guess introducing the default case that Audrius proposed can't cause any harm (it did solve your problem, right?), but I am still confused about how this can happen.

@Zillode Any inputs on this?

**Comment by ctxnop** [comment]
Yes I agree that the flow should not bet concurrent, but the context is still a concurrent one, and the documentation said you cannot do this in a concurrent.

The system of go routine seem to be a lot like the async/await system of the .Net which I know better.
Maybe it's different, you'll tell me, but if not, here what happen if you write an equivalent code:
- There is an implicit hoare monitor
- Flow is splitted on subtask
- A scheduler execute tasks like he want, not like you think he will, using the hoare monitor to define which task should be run before/after/concurrently with other tasks..
- It can either use a message pump or a thread pool, depending on some criteria, to run the whole thing.

Of course, the semantic should not be changed by the scheduler, but you cannot control where the mutexes, the tasks, threads, etc... appears. (In fact, you can control those things, but then you will write a lot more code with explicit concurrent framework call, so you lose the main interest which is a simple syntax to write efficient concurrent code)

So this thing may happen:
1) a task A is created for the "if" block in the beginning of the "for" block.
2) a task B, C and D are created for each "case" of the select.
3) a task E is created for the closing brace of the select.
4) B, C and D are configured to wait for the result of A.
5) E is configured to wait for the result of any task in B, C and D.

Then, when B, is done, E start, but C and D may still be alive before E is done, so when A is run again, it is run concurrently with C and D. So that D may read the timer channel in the same time that A. In which case A will block because data is already read by D, and since the timer isn't reset, nothing will ever occurs again on this channel.

As I said, it may be not the case in go, I don't know.

About the default case, I'm not sure it solved my problem, because as I said, I misplaced it at first.
Then I moved it to what I'm guessing is the right place, but I didn't test long enough to be sure.
So not sure if the problem is really gone, nor if I placed it in the right place (I put it right before the select's closing brace).
More important, I didn't check but I guess it consume a lot a CPU doing this, because it basically is a "while(true) {}" the whole time nothing occurs on the 3 other cases, which is almost 99% of the time.
Am I right ?

So, to me, it can't be a solution.

I cannot post log right now nor test more over the default case (I'm not on the right computer). But I will later if needed.

**Comment by calmh** [comment]
The question about concurrency here is if some *other* goroutine can receive from `flushTimer.C` after the call to `flushTimer.Stop()` and the receive in

```go
	if !flushTimer.Stop() {
		<-flushTimer.C
```

If so, it needs guarding by the select as mentioned. If there there is no other goroutine receiving from `flushTimer.C` then this is not the issue.

**Comment by imsodin** [comment]
There is no other goroutine which can receive from `flushTimer.C`. It only appears within `accumulateChanges` (which is only one run once as a goroutine) and `flushTimer` is never passed to another function.
That's what is weird here: There actually cannot be such a problem, but according to ctxnop's investigations, there is.

**Comment by ctxnop** [comment]
Once again, I known nothing about go, so I may be completely wrong, missed something, etc...
But what I'm sure about is, if I apply my patch, then I don't have problem anymore, so I'm pretty confident that this timer play a role in the issue.
I will try again tomorrow or Monday and keep you in touch.

**Comment by ctxnop** [comment]
I just tried to make a [short, self-contained, correct example](http://sscce.org/) to see if I reproduce the problem. I see no major difference between my example and the original code, and it kinda do what I expected, but instead of freezing, it crash with this log:
```
C:\dev\projects\go_workspace\src\github.com\ctxnop\timerpoc>go build && timerpoc.exe
Before reset test
Before !timer.Stop() test
Before <-timer.C wait
Before select wait
Inside case <-timer.C
Before reset test
Before !timer.Stop() test
Before <-timer.C wait
fatal error: all goroutines are asleep - deadlock!

goroutine 1 [chan receive]:
main.main()
        C:/dev/projects/go_workspace/src/github.com/ctxnop/timerpoc/timerpoc.go:45 +0xea

goroutine 18 [chan receive]:
main.test(0xc042042060, 0xc0420420c0, 0xc042042120, 0xc0420441c0)
        C:/dev/projects/go_workspace/src/github.com/ctxnop/timerpoc/timerpoc.go:19 +0x621
created by main.main
        C:/dev/projects/go_workspace/src/github.com/ctxnop/timerpoc/timerpoc.go:43 +0xbd
```

Here my example, maybe it does not correspond to the original code, you'll tell me:
```go
package main

import (
	"fmt"
	"os"
	"time"
)

func test(chanCode chan int, chanInput1 chan string, chanInput2 chan string) func(string) {
	timer := time.NewTimer(0)
	reset := true
	for {
		fmt.Println("Before reset test")
		if reset {
			reset = false
			fmt.Println("Before !timer.Stop() test")
			if !timer.Stop() {
				fmt.Println("Before <-timer.C wait")
				<-timer.C
			}
			timer.Reset(100)
		}
		fmt.Println("Before select wait")
		select {
		case item := <-chanInput1:
			fmt.Println("Inside case <-chan_input1 with item=" + item)
			reset = true
		case item := <-chanInput2:
			fmt.Println("Inside case <-chan_input2 with item=" + item)
			reset = true
		case <-timer.C:
			fmt.Println("Inside case <-timer.C")
			reset = true
		}
	}
}

func main() {
	code := make(chan int)
	input1 := make(chan string)
	input2 := make(chan string)

	go test(code, input1, input2)

	retCode := <-code
	os.Exit(retCode)
}
```
I'm sure this example can be simplified again, if I was knowing what I was writing...

**Comment by ctxnop** [comment]
So, I didn't want to wait to get onto my other computer, so I installed everything on another one, still Windows 10.

here is my `accumulateChanges` function:
```go
func accumulateChanges(debounceTimeout time.Duration,
	folder string,
	folderPath string,
	dirVsFiles int,
	stInput chan STEvent,
	fsInput chan string,
	callback InformCallback) func(string) {
	var delayScanInterval time.Duration
	if delayScan > 0 {
		delayScanInterval = time.Duration(delayScan-5) * time.Second
		Debug.Printf("Delay scan reminder interval for %s set to %.0f seconds\n", folder, delayScanInterval.Seconds())
	} else {
		// If delayScan is set to 0, then we never send requests to delay full scans.
		// "9999 * time.Hour" here is an approximation of "forever".
		delayScanInterval = 9999 * time.Hour
		Debug.Println("Delay scan reminders are disabled")
	}
	inProgress := make(map[string]progressTime) // [path string]{fs, start}
	currInterval := delayScanInterval           // Timeout of the timer
	if delayScan > 0 {
		askToDelayScan(folder, callback)
	}
	nextScanTime := time.Now().Add(delayScanInterval) // Time to remind Syncthing to delay scan
	flushTimer := time.NewTimer(0)
	flushTimerNeedsReset := true
	for {
		Debug.Println("[ISSUE] Inside the 'for' loop.")
		if flushTimerNeedsReset {
			flushTimerNeedsReset = false
			Debug.Println("[ISSUE] Just before the 'if !flushTimer.Stop()'.")
			if !flushTimer.Stop() {
				Debug.Println("[ISSUE] Just before the '<-flushTimer.C'.")
				<-flushTimer.C
				Debug.Println("[ISSUE] Just after the '<-flushTimer.C'.")
			}
			flushTimer.Reset(currInterval)
		}
		Debug.Println("[ISSUE] Just before the 'select'.")
		select {
		case item := <-stInput:
			Debug.Println("[ISSUE] Inside the 'case item := <-stInput'.")
			// No other change
		case item := <-fsInput:
			Debug.Println("[ISSUE] Inside the 'case item := <-fsInput'.")
			// No other change
		case <-flushTimer.C:
			Debug.Println("[ISSUE] Inside the '<-flushTimer.C'.")
			// No other change
			if err == nil {
				nextScanTime = time.Now().Add(delayScanInterval) // Scan was delayed
			} else {
				Warning.Println("Syncthing failed to index changes for ", folder, err)
				time.Sleep(configSyncTimeout)
			}
			//default:
		}
	}
}
```

Here is the console ouput:
```
C:\dev\projects\go_workspace\src\github.com\syncthing\syncthing-inotify>go build && syncthing-inotify.exe -verbosity=4
[TRACE] 22:49:10 Testing WebGUI
[TRACE] 22:49:10 Getting Folders
[DEBUG] 22:49:10 Installing watch for Default Folder
[TRACE] 22:49:10 Requesting STEvents: 0
[TRACE] 22:49:10 Getting ignore patterns for Default Folder
[OK] 22:49:10 Watching Default Folder: C:\Users\ctxnop\Sync
[DEBUG] 22:49:10 Delay scan reminder interval for default set to 3595 seconds
[TRACE] 22:49:10 Asking to delay full scanning of default
[TRACE] 22:49:10 Informing ST: default: [.stfolder]
[TRACE] 22:49:10 ConfigSaved, exiting if folders changed
[TRACE] 22:49:10 ConfigSaved, exiting if folders changed
[OK] 22:49:10 Syncthing is indexing change in default: [.stfolder]
[TRACE] 22:49:10 ConfigSaved, exiting if folders changed
[DEBUG] 22:49:10 [ISSUE] Inside the 'for' loop.
[TRACE] 22:49:10 ConfigSaved, exiting if folders changed
[DEBUG] 22:49:10 [ISSUE] Just before the 'if !flushTimer.Stop()'.
[TRACE] 22:49:10 Requesting STEvents: 30
[DEBUG] 22:49:10 [ISSUE] Just before the '<-flushTimer.C'.
[DEBUG] 22:49:10 [ISSUE] Just after the '<-flushTimer.C'.
[DEBUG] 22:49:10 [ISSUE] Just before the 'select'.
[TRACE] 22:49:10 Waiting for Sync
[TRACE] 22:49:10 Waiting for Sync
[TRACE] 22:49:10 Waiting for Sync
[TRACE] 22:49:10 Waiting for Sync
[TRACE] 22:49:10 Requesting STEvents: 32
[TRACE] 22:49:10 Getting Folders
[TRACE] 22:49:10 Getting Folders
[TRACE] 22:49:10 Getting Folders
[TRACE] 22:49:10 Getting Folders
[DEBUG] 22:49:29 Change detected in: C:\Users\ctxnop\Sync\new_file.txt (could still be ignored)
[TRACE] 22:49:29 Change detected in: C:\Users\ctxnop\Sync\new_file.txt
[DEBUG] 22:49:29 [ISSUE] Inside the 'case item := <-fsInput'.
[DEBUG] 22:49:29 [FS] Incoming Changes for default, speeding up inotify timeout parameters to 500ms
[DEBUG] 22:49:29 [FS] Tracking: new_file.txt
[DEBUG] 22:49:29 [ISSUE] Inside the 'for' loop.
[DEBUG] 22:49:29 [ISSUE] Just before the 'if !flushTimer.Stop()'.
[DEBUG] 22:49:29 [ISSUE] Just before the 'select'.
[DEBUG] 22:49:30 [ISSUE] Inside the '<-flushTimer.C'.
[DEBUG] 22:49:30 Timeout AccumulateChanges
[DEBUG] 22:49:30 Informing about new_file.txt
[DEBUG] 22:49:30 [AG] Not found: new_file.txt
[DEBUG] 22:49:30 [AG] Appending path: new_file.txt new_file.txt
[TRACE] 22:49:30 Informing ST: default: [new_file.txt]
[TRACE] 22:49:30 Requesting STEvents: 33
[OK] 22:49:30 Syncthing is indexing change in default: [new_file.txt]
[DEBUG] 22:49:30 [INFORMED] Removed tracking for new_file.txt
[DEBUG] 22:49:30 [ISSUE] Inside the 'for' loop.
[DEBUG] 22:49:30 [ISSUE] Just before the 'if !flushTimer.Stop()'.
[DEBUG] 22:49:30 [ISSUE] Just before the '<-flushTimer.C'.
[TRACE] 22:49:30 Requesting STEvents: 35
[TRACE] 22:49:31 Requesting STEvents: 36
[DEBUG] 22:49:47 Change detected in: C:\Users\ctxnop\Sync\new_file.txt (could still be ignored)
[TRACE] 22:49:47 Change detected in: C:\Users\ctxnop\Sync\new_file.txt
[TRACE] 22:50:31 Requesting STEvents: 36
[TRACE] 22:51:31 Requesting STEvents: 36
[TRACE] 22:52:31 Requesting STEvents: 36
[TRACE] 22:53:31 Requesting STEvents: 36
[TRACE] 22:54:31 Requesting STEvents: 36
[TRACE] 22:55:31 Requesting STEvents: 36
[TRACE] 22:56:31 Requesting STEvents: 36
```
I created a file 'new_file.txt'  at 22:49:30 and removed at 22:49:47. Then I created and removed a file 'another_file.txt' at 22:50. I let it run til 23:00 but nothing happen, changes are still not detected anymore.

But there is something I just noticed: if you look at the 22:49:47 log, it's from the watchFolder function. Any change since this log is triggered do not produce any other log. So the watchFolder function is blocked too.

Now, you may notice I commented out the "default" case in the select. Here what happen if I uncomment it:
Just like I guessed the CPU is burning one of it's core with a beautiful 99% usage, but then something unexpected happen when I created a file:
```
[DEBUG] 23:12:35 [ISSUE] Just before the 'select'.
[DEBUG] 23:12:35 [ISSUE] Inside the 'for' loop.
[DEBUG] 23:12:35 [ISSUE] Just before the 'select'.
[DEBUG] 23:12:35 [ISSUE] Inside the 'for' loop.
[DEBUG] 23:12:35 [ISSUE] Just before the 'select'.
[DEBUG] 23:12:35 [ISSUE] Inside the 'for' loop.
[DEBUG] 23:12:35 [ISSUE] Just before the 'select'.
[DEBUG] 23:12:35 [ISSUE] Inside the '<-flushTimer.C'.
[DEBUG] 23:12:35 Timeout AccumulateChanges
[DEBUG] 23:12:35 Informing about test.txt
[DEBUG] 23:12:35 [AG] Not found: test.txt
[DEBUG] 23:12:35 [AG] Appending path: test.txt test.txt
[TRACE] 23:12:35 Informing ST: default: [test.txt]
[TRACE] 23:12:35 Requesting STEvents: 45
[OK] 23:12:35 Syncthing is indexing change in default: [test.txt]
[DEBUG] 23:12:35 [INFORMED] Removed tracking for test.txt
[DEBUG] 23:12:35 [ISSUE] Inside the 'for' loop.
[DEBUG] 23:12:35 [ISSUE] Just before the 'if !flushTimer.Stop()'.
[DEBUG] 23:12:35 [ISSUE] Just before the '<-flushTimer.C'.
[TRACE] 23:12:35 Requesting STEvents: 47
[TRACE] 23:12:35 Requesting STEvents: 48
[DEBUG] 23:12:45 Change detected in: C:\Users\ctxnop\Sync\other_test.txt (could still be ignored)
[TRACE] 23:12:45 Change detected in: C:\Users\ctxnop\Sync\other_test.txt
```
So, the two first line were repeating very fast, as expected, but the "[ISSUE] Inside the '<-flushTimer.C'." occurs when I created a file. Then, everything goes back to normal, as if I didn't uncommented the default case. That is weird to me.
But the conclusion is that the default case did not solve my problem.

Here is what happen is I just comment out the code like this (and remove the default case):
```go
	for {
		Debug.Println("[ISSUE] Inside the 'for' loop.")
		if flushTimerNeedsReset {
			flushTimerNeedsReset = false
			Debug.Println("[ISSUE] Just before the 'if !flushTimer.Stop()'.")
			//if !flushTimer.Stop() {
			//	Debug.Println("[ISSUE] Just before the '<-flushTimer.C'.")
			//	<-flushTimer.C
			//	Debug.Println("[ISSUE] Just after the '<-flushTimer.C'.")
			//}
			flushTimer.Reset(currInterval)
		}
```
```
C:\dev\projects\go_workspace\src\github.com\syncthing\syncthing-inotify>go build && syncthing-inotify.exe -verbosity=4
[TRACE] 23:19:17 Testing WebGUI
[TRACE] 23:19:17 Getting Folders
[DEBUG] 23:19:17 Installing watch for Default Folder
[TRACE] 23:19:17 Requesting STEvents: 0
[TRACE] 23:19:17 Getting ignore patterns for Default Folder
[OK] 23:19:17 Watching Default Folder: C:\Users\ctxnop\Sync
[DEBUG] 23:19:17 Delay scan reminder interval for default set to 3595 seconds
[TRACE] 23:19:17 Asking to delay full scanning of default
[TRACE] 23:19:17 Informing ST: default: [.stfolder]
[OK] 23:19:17 Syncthing is indexing change in default: [.stfolder]
[DEBUG] 23:19:17 [ISSUE] Inside the 'for' loop.
[DEBUG] 23:19:17 [ISSUE] Just before the 'if !flushTimer.Stop()'.
[DEBUG] 23:19:17 [ISSUE] Just before the 'select'.
[TRACE] 23:19:17 ConfigSaved, exiting if folders changed
[TRACE] 23:19:17 ConfigSaved, exiting if folders changed
[TRACE] 23:19:17 ConfigSaved, exiting if folders changed
[TRACE] 23:19:17 Waiting for Sync
[TRACE] 23:19:17 Waiting for Sync
[TRACE] 23:19:17 ConfigSaved, exiting if folders changed
[TRACE] 23:19:17 Requesting STEvents: 52
[TRACE] 23:19:17 Waiting for Sync
[TRACE] 23:19:17 Waiting for Sync
[TRACE] 23:19:17 Getting Folders
[TRACE] 23:19:17 Requesting STEvents: 54
[TRACE] 23:19:17 Getting Folders
[TRACE] 23:19:17 Getting Folders
[TRACE] 23:19:17 Getting Folders
[DEBUG] 23:19:27 Change detected in: C:\Users\ctxnop\Sync\test1.txt (could still be ignored)
[TRACE] 23:19:27 Change detected in: C:\Users\ctxnop\Sync\test1.txt
[DEBUG] 23:19:27 [ISSUE] Inside the 'case item := <-fsInput'.
[DEBUG] 23:19:27 [FS] Incoming Changes for default, speeding up inotify timeout parameters to 500ms
[DEBUG] 23:19:27 [FS] Tracking: test1.txt
[DEBUG] 23:19:27 [ISSUE] Inside the 'for' loop.
[DEBUG] 23:19:27 [ISSUE] Just before the 'if !flushTimer.Stop()'.
[DEBUG] 23:19:27 [ISSUE] Just before the 'select'.
[DEBUG] 23:19:27 [ISSUE] Inside the '<-flushTimer.C'.
[DEBUG] 23:19:27 Timeout AccumulateChanges
[DEBUG] 23:19:27 Informing about test1.txt
[DEBUG] 23:19:27 [AG] Not found: test1.txt
[DEBUG] 23:19:27 [AG] Appending path: test1.txt test1.txt
[TRACE] 23:19:27 Informing ST: default: [test1.txt]
[TRACE] 23:19:27 Requesting STEvents: 55
[OK] 23:19:27 Syncthing is indexing change in default: [test1.txt]
[DEBUG] 23:19:27 [INFORMED] Removed tracking for test1.txt
[DEBUG] 23:19:27 [ISSUE] Inside the 'for' loop.
[DEBUG] 23:19:27 [ISSUE] Just before the 'if !flushTimer.Stop()'.
[DEBUG] 23:19:27 [ISSUE] Just before the 'select'.
[TRACE] 23:19:27 Requesting STEvents: 57
[DEBUG] 23:19:28 [ISSUE] Inside the '<-flushTimer.C'.
[DEBUG] 23:19:28 Slowing down inotify timeout parameters for default
[DEBUG] 23:19:28 [ISSUE] Inside the 'for' loop.
[DEBUG] 23:19:28 [ISSUE] Just before the 'if !flushTimer.Stop()'.
[DEBUG] 23:19:28 [ISSUE] Just before the 'select'.
[TRACE] 23:19:29 Requesting STEvents: 58
[DEBUG] 23:19:41 Change detected in: C:\Users\ctxnop\Sync\test2.txt (could still be ignored)
[TRACE] 23:19:41 Change detected in: C:\Users\ctxnop\Sync\test2.txt
[DEBUG] 23:19:41 [ISSUE] Inside the 'case item := <-fsInput'.
[DEBUG] 23:19:41 [FS] Incoming Changes for default, speeding up inotify timeout parameters to 500ms
[DEBUG] 23:19:41 [FS] Tracking: test2.txt
[DEBUG] 23:19:41 [ISSUE] Inside the 'for' loop.
[DEBUG] 23:19:41 [ISSUE] Just before the 'if !flushTimer.Stop()'.
[DEBUG] 23:19:41 [ISSUE] Just before the 'select'.
[DEBUG] 23:19:41 [ISSUE] Inside the '<-flushTimer.C'.
[DEBUG] 23:19:41 Timeout AccumulateChanges
[DEBUG] 23:19:41 Informing about test2.txt
[DEBUG] 23:19:41 [AG] Not found: test2.txt
[DEBUG] 23:19:41 [AG] Appending path: test2.txt test2.txt
[TRACE] 23:19:41 Informing ST: default: [test2.txt]
[TRACE] 23:19:41 Requesting STEvents: 59
[OK] 23:19:41 Syncthing is indexing change in default: [test2.txt]
[DEBUG] 23:19:41 [INFORMED] Removed tracking for test2.txt
[DEBUG] 23:19:41 [ISSUE] Inside the 'for' loop.
[DEBUG] 23:19:41 [ISSUE] Just before the 'if !flushTimer.Stop()'.
[DEBUG] 23:19:41 [ISSUE] Just before the 'select'.
[TRACE] 23:19:41 Requesting STEvents: 61
[TRACE] 23:19:41 Requesting STEvents: 62
[DEBUG] 23:19:42 [ISSUE] Inside the '<-flushTimer.C'.
[DEBUG] 23:19:42 Slowing down inotify timeout parameters for default
[DEBUG] 23:19:42 [ISSUE] Inside the 'for' loop.
[DEBUG] 23:19:42 [ISSUE] Just before the 'if !flushTimer.Stop()'.
[DEBUG] 23:19:42 [ISSUE] Just before the 'select'.
[DEBUG] 23:19:51 Change detected in: C:\Users\ctxnop\Sync\test3.txt (could still be ignored)
[TRACE] 23:19:51 Change detected in: C:\Users\ctxnop\Sync\test3.txt
[DEBUG] 23:19:51 [ISSUE] Inside the 'case item := <-fsInput'.
[DEBUG] 23:19:51 [FS] Incoming Changes for default, speeding up inotify timeout parameters to 500ms
[DEBUG] 23:19:51 [FS] Tracking: test3.txt
[DEBUG] 23:19:51 [ISSUE] Inside the 'for' loop.
[DEBUG] 23:19:51 [ISSUE] Just before the 'if !flushTimer.Stop()'.
[DEBUG] 23:19:51 [ISSUE] Just before the 'select'.
[DEBUG] 23:19:51 [ISSUE] Inside the '<-flushTimer.C'.
[DEBUG] 23:19:51 Timeout AccumulateChanges
[DEBUG] 23:19:51 Informing about test3.txt
[DEBUG] 23:19:51 [AG] Not found: test3.txt
[DEBUG] 23:19:51 [AG] Appending path: test3.txt test3.txt
[TRACE] 23:19:51 Informing ST: default: [test3.txt]
[TRACE] 23:19:51 Requesting STEvents: 63
[OK] 23:19:51 Syncthing is indexing change in default: [test3.txt]
[DEBUG] 23:19:51 [INFORMED] Removed tracking for test3.txt
[DEBUG] 23:19:51 [ISSUE] Inside the 'for' loop.
[DEBUG] 23:19:51 [ISSUE] Just before the 'if !flushTimer.Stop()'.
[DEBUG] 23:19:51 [ISSUE] Just before the 'select'.
[TRACE] 23:19:51 Requesting STEvents: 65
[TRACE] 23:19:51 Requesting STEvents: 66
[DEBUG] 23:19:52 [ISSUE] Inside the '<-flushTimer.C'.
[DEBUG] 23:19:52 Slowing down inotify timeout parameters for default
[DEBUG] 23:19:52 [ISSUE] Inside the 'for' loop.
[DEBUG] 23:19:52 [ISSUE] Just before the 'if !flushTimer.Stop()'.
[DEBUG] 23:19:52 [ISSUE] Just before the 'select'.
```
So, as you see, I'm now able to create 3 files (and more) without issues, and no CPU usage.

With all these log and code, my comments begins to look like a book... Sorry about that.

**Comment by imsodin** [comment]
> With all these log and code, my comments begins to look like a book... Sorry about that.

I must admit after reading your previous post, I only skimmed over that one. Thanks anyone for this comprehensive reporting and testing. I must admit, at times I couldn't follow.

So as far as I understand the problem is simple and has nothing to do with concurrency:
`flushtimer.C` is drained on reset, even though it might already be drained at that point (`notifyTimerNeedsReset` is set to true in other places than the `case <-flushrimer.C` part.

The new riddle: How on earth did and does this work on many systems?

**Comment by ctxnop** [comment]
So, as I saw your patch, it mean that I completly misunderstood the proposition made by @AudriusButkevicius 
I tried your patch and it work like a charm. I consider this issue closed then, thanks everyone.

**Comment by Sebastien-D** [comment]
Would you please tell me why fix #164 / pull request #165 seems not merged into master branch ?
I also need it.
Thanks.
Seb.

**Comment by imsodin** [comment]
It *was* merged to master. My local master branch has diverged from the github one, it looks like there was some kind of rebasing going on on the 7.5.
@Zillode: Do you know what was going on? (And do you know who lhoste-bell is?)

**Comment by calmh** [comment]
There was some rebase damage. I think we've recovered it now.

**Comment by Sebastien-D** [comment]
Thanks a lot.
Do you know when the next release (probably v0.8.6) is planned ?
IOW when this fix will be integrated in a release ?
Thanks.

**Comment by Zillode** [comment]
I plan to release a new version this weekend.

**Comment by Sebastien-D** [comment]
Great !
Have a good weekend.

**Comment by Sebastien-D** [comment]
I still not see any new release :-(
I would really appreciate to have a new release that integrates #164 fix.
Thanks

**Comment by Zillode** [comment]
I really wanted to test this on my macOS machine first. Looks good, thanks for the help @imsodin
